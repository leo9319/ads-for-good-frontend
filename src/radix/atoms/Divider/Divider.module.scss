@import '../../../styles/colours';

.divider {
  margin: 0;
  padding: 0;

  // Horizontal divider styles
  &-horizontal {
    height: 1px;
  }

  // Vertical divider styles
  &-vertical {
    width: 1px;
  }

  // Size 1 (16px)
  &-size-1 {
    width: 16px;
    height: 16px;
  }

  // Size 2 (32px)
  &-size-2 {
    width: 32px;
    height: 32px;
  }

  // Size 3 (64px)
  &-size-3 {
    width: 64px;
    height: 64px;
  }

  // Color variants
  &-primary {
    background-color: $color-neutral-300;
  }

  &-secondary {
    background-color: $color-neutral-100;
  }

  &-inverse {
    background-color: $color-white;
  }
}
