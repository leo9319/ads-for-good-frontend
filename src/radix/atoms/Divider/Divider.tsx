import React from 'react';
import classNames from 'classnames';

import styles from './Divider.module.scss';

/**
 * Props for the Divider component.
 *
 * @property orientation - Direction of the divider: "horizontal" (default) or "vertical".
 * @property color - Color variant of the divider: "primary", "secondary", or "inverse".
 * @property size - Size of the divider:
 *    1 = 16px,
 *    2 = 32px,
 *    3 = 64px,
 *    4 = fill container (100% width or height depending on orientation).
 */

export type DividerProps = {
  orientation?: 'horizontal' | 'vertical';
  color?: 'primary' | 'secondary' | 'inverse';
  size?: 1 | 2 | 3 | 4; // size: 4 means fill container
};

export const Divider: React.FC<DividerProps> = ({
  orientation = 'horizontal',
  color = 'primary',
  size = 1,
}) => {
  const validOrientations = ['horizontal', 'vertical'];
  const validColors = ['primary', 'secondary', 'inverse'];
  const validSizes = [1, 2, 3, 4];

  const safeOrientation = validOrientations.includes(orientation)
    ? orientation
    : 'horizontal';
  const safeColor = validColors.includes(color) ? color : 'primary';
  const safeSize = validSizes.includes(size) ? size : 1;

  const sizeMap = {
    1: 16,
    2: 32,
    3: 64,
  };

  const isFill = safeSize === 4;

  const dividerClass = classNames(styles.divider, 'divider', {
    [styles[`divider-${safeOrientation}`]]: true,
    [styles[`divider-${safeColor}`]]: true,
    [styles[`divider-size-${safeSize}`]]: !isFill,
    [styles['divider-fill']]: isFill,
  });

  const style: React.CSSProperties = isFill
    ? safeOrientation === 'horizontal'
      ? { width: '100%', minWidth: '128px', height: '1px' }
      : { height: '100%', minHeight: '128px', width: '1px' }
    : safeOrientation === 'horizontal'
      ? { width: sizeMap[safeSize as 1 | 2 | 3], height: '1px' }
      : { height: sizeMap[safeSize as 1 | 2 | 3], width: '1px' };

  return <div role="separator" className={dividerClass} style={style} />;
};

export default Divider;
