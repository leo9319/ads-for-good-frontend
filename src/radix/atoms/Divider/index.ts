export type { DividerProps } from './Divider';
export { Divider } from './Divider';
export { default } from './Divider';
