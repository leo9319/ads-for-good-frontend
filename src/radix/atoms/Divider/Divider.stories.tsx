import type { Meta, StoryObj } from '@storybook/react';
import React from 'react';
import { Divider, DividerProps } from './Divider';

const meta: Meta<typeof Divider> = {
  title: 'Library/Atoms/Divider',
  component: Divider,
  tags: ['autodocs'],
  argTypes: {
    orientation: {
      control: 'radio',
      options: ['horizontal', 'vertical'],
      defaultValue: 'horizontal', // Explicit default value for orientation
    },
    color: {
      control: 'select',
      options: ['primary', 'secondary', 'inverse'],
      defaultValue: 'primary', // Explicit default value for color
    },
    size: {
      control: 'radio',
      options: [1, 2, 3, 4],
      defaultValue: 1, // Explicit default value for size
    },
  },
};

export default meta;

type Story = StoryObj<DividerProps>;

// Reusable horizontal container
const Container = ({ children }: { children: React.ReactNode }) => (
  <div
    style={{
      width: '300px',
      border: '1px dashed #ccc',
      padding: '16px',
      marginBottom: '24px',
    }}
  >
    {children}
  </div>
);

// Reusable vertical container
const VerticalContainer = ({ children }: { children: React.ReactNode }) => (
  <div
    style={{
      height: '150px',
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
      border: '1px dashed #ccc',
      padding: '16px',
    }}
  >
    {children}
  </div>
);

// Default horizontal example
export const Horizontal: Story = {
  args: {
    orientation: 'horizontal',
    color: 'primary',
    size: 1,
  },
};

// Default vertical example
export const Vertical: Story = {
  args: {
    orientation: 'vertical',
    color: 'primary',
    size: 2,
  },
};

// Size 4 divider inside a container (should fill the container width)
export const Size4HorizontalFillInContainer: Story = {
  render: args => (
    <Container>
      <Divider {...args} />
    </Container>
  ),
  args: {
    orientation: 'horizontal',
    color: 'primary',
    size: 4,
  },
};

// Size 4 vertical divider inside a vertical container (should fill the container height)
export const Size4VerticalFillInContainer: Story = {
  render: args => (
    <VerticalContainer>
      <Divider {...args} />
    </VerticalContainer>
  ),
  args: {
    orientation: 'vertical',
    color: 'secondary',
    size: 4,
  },
};
