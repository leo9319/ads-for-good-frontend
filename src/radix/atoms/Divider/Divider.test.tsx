import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Divider } from './Divider';

describe('Divider Component', () => {
  it('renders a horizontal divider with default properties', () => {
    render(<Divider orientation="horizontal" />);
    const divider = screen.getByRole('separator');
    expect(divider).toHaveClass('divider-horizontal');
    expect(divider).toHaveClass('divider-size-1');
    expect(divider).toHaveStyle('height: 1px');
    expect(divider).toHaveStyle('width: 16px');
  });

  it('renders a vertical divider with default properties', () => {
    render(<Divider orientation="vertical" />);
    const divider = screen.getByRole('separator');
    expect(divider).toHaveClass('divider-vertical');
    expect(divider).toHaveClass('divider-size-1');
    expect(divider).toHaveStyle('width: 1px');
    expect(divider).toHaveStyle('height: 16px');
  });

  it('renders a vertical divider with size 3 (64px height)', () => {
    render(<Divider orientation="vertical" size={3} />);
    const divider = screen.getByRole('separator');
    expect(divider).toHaveStyle('height: 64px');
    expect(divider).toHaveClass('divider-size-3');
  });

  it('renders a horizontal divider with size 2 (32px width)', () => {
    render(<Divider orientation="horizontal" size={2} />);
    const divider = screen.getByRole('separator');
    expect(divider).toHaveStyle('width: 32px');
    expect(divider).toHaveClass('divider-size-2');
  });

  it('renders a horizontal divider with size 4 (fills container)', () => {
    render(<Divider orientation="horizontal" size={4} />);
    const divider = screen.getByRole('separator');
    expect(divider).toHaveClass('divider-fill');
    expect(divider).toHaveStyle('width: 100%');
    expect(divider).toHaveStyle('min-width: 128px');
  });

  it('renders a vertical divider with size 4 (fills container)', () => {
    render(<Divider orientation="vertical" size={4} />);
    const divider = screen.getByRole('separator');
    expect(divider).toHaveClass('divider-fill');
    expect(divider).toHaveStyle('height: 100%');
    expect(divider).toHaveStyle('min-height: 128px');
  });

  it('applies the correct color (primary)', () => {
    render(<Divider color="primary" />);
    const divider = screen.getByRole('separator');
    expect(divider).toHaveClass('divider-primary');
  });

  it('applies the correct color (inverse)', () => {
    render(<Divider color="inverse" />);
    const divider = screen.getByRole('separator');
    expect(divider).toHaveClass('divider-inverse');
  });

  it('renders with default props when none are provided', () => {
    render(<Divider />);
    const divider = screen.getByRole('separator');
    expect(divider).toHaveClass('divider-horizontal');
    expect(divider).toHaveClass('divider-primary');
    expect(divider).toHaveStyle('width: 16px');
    expect(divider).toHaveStyle('height: 1px');
  });

  it('falls back to default values when given invalid props', () => {
    // @ts-expect-error â€“ intentionally incorrect props
    render(<Divider orientation="diagonal" size={99} color="rainbow" />);
    const divider = screen.getByRole('separator');

    expect(divider).toHaveClass('divider-horizontal');
    expect(divider).toHaveClass('divider-primary');
    expect(divider).toHaveClass('divider-size-1');
    expect(divider).toHaveStyle('width: 16px');
    expect(divider).toHaveStyle('height: 1px');
  });

  it('has the correct ARIA role', () => {
    render(<Divider />);
    const divider = screen.getByRole('separator');
    expect(divider).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { container } = render(<Divider />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
