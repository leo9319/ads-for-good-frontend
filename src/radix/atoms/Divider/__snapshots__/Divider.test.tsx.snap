// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Divider Component matches snapshot 1`] = `
<div
  class="divider divider divider-horizontal divider-primary divider-size-1"
  role="separator"
  style="width: 16px; height: 1px;"
/>
`;
